src/witness_generator/machines/machine_extractor.rs

```diff
@@ -1,29 +1,33 @@
+use std::collections::BTreeMap;
 use std::collections::HashSet;
 
+use super::block_machine::BlockMachine;
 use super::double_sorted_witness_machine::DoubleSortedWitnesses;
 use super::fixed_lookup_machine::FixedLookup;
 use super::sorted_witness_machine::SortedWitnesses;
 use super::FixedData;
 use super::Machine;
 use crate::analyzer::{Expression, Identity, SelectedExpressions};
+use crate::witness_generator::bit_constraints::BitConstraint;
 use crate::witness_generator::WitnessColumn;
 
 /// Finds machines in the witness columns and identities
 /// and returns a list of machines and the identities
 /// that are not "internal" to the machines.
 pub fn split_out_machines<'a>(
     fixed: &'a FixedData<'a>,
-    identities: &'a [Identity],
+    identities: Vec<&'a Identity>,
     witness_cols: &'a [WitnessColumn],
+    global_bit_constraints: &BTreeMap<&'a str, BitConstraint>,
 ) -> (FixedLookup, Vec<Box<dyn Machine>>, Vec<&'a Identity>) {
     let fixed_lookup = FixedLookup::try_new(fixed, &[], &Default::default()).unwrap();
 
     let mut machines: Vec<Box<dyn Machine>> = vec![];
 
     let all_witnesses = witness_cols.iter().map(|c| c.name).collect::<HashSet<_>>();
     let mut remaining_witnesses = all_witnesses.clone();
-    let mut base_identities = identities.iter().collect::<Vec<_>>();
-    for id in identities {
+    let mut base_identities = identities.to_vec();
```

Comment:
Isn't it a Vec already?

Reply:
Yes, but we want a copy. Maybe clone would also work.

Reply:
Both work, clone is slightly clearer